File: leetcode/Daily Question/javascript/2092. Find All People With Secret.js
/*
https://leetcode.com/problems/find-all-people-with-secret/?envType=daily-question&envId=2025-12-19

Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/
*/

var findAllPeople = function(n, meetings, firstPerson) {
    meetings.sort((a, b) => a[2] - b[2]);

    const knows = Array(n).fill(false);
    knows[0] = true;
    knows[firstPerson] = true;

    let i = 0;
    while (i < meetings.length) {
        const t = meetings[i][2];
        const adj = new Map();
        const participants = new Set();

        while (i < meetings.length && meetings[i][2] === t) {
            const [x, y] = meetings[i];
            if (!adj.has(x)) adj.set(x, []);
            if (!adj.has(y)) adj.set(y, []);
            adj.get(x).push(y);
            adj.get(y).push(x);
            participants.add(x);
            participants.add(y);
            i++;
        }

        const queue = [];
        const visited = new Set();
        for (const p of participants) {
            if (knows[p]) {
                queue.push(p);
                visited.add(p);
            }
        }

        while (queue.length) {
            const cur = queue.shift();
            if (!adj.has(cur)) continue;
            for (const nxt of adj.get(cur)) {
                if (!visited.has(nxt)) {
                    visited.add(nxt);
                    queue.push(nxt);
                }
            }
        }

        for (const p of visited) {
            knows[p] = true;
        }
    }

    return knows.map((v, idx) => v ? idx : -1).filter(x => x !== -1);
};

File: leetcode/Daily Question/csharp/2092. Find All People With Secret.cs
/*
https://leetcode.com/problems/find-all-people-with-secret/?envType=daily-question&envId=2025-12-19

Автор: Дуплей Максим Игоревич - AGLA
ORCID: https://orcid.org/0009-0007-7605-539X
GitHub: https://github.com/QuadDarv1ne/
*/

public class Solution {
    public IList<int> FindAllPeople(int n, int[][] meetings, int firstPerson) {
        Array.Sort(meetings, (a, b) => a[2].CompareTo(b[2]));

        bool[] knows = new bool[n];
        knows[0] = true;
        knows[firstPerson] = true;

        int i = 0, m = meetings.Length;
        while (i < m) {
            int currentTime = meetings[i][2];
            var adj = new Dictionary<int, List<int>>();
            var participants = new HashSet<int>();

            while (i < m && meetings[i][2] == currentTime) {
                int x = meetings[i][0], y = meetings[i][1];
                if (!adj.ContainsKey(x)) adj[x] = new List<int>();
                if (!adj.ContainsKey(y)) adj[y] = new List<int>();
                adj[x].Add(y);
                adj[y].Add(x);
                participants.Add(x);
                participants.Add(y);
                i++;
            }

            var queue = new Queue<int>();
            var visited = new HashSet<int>();
            foreach (int p in participants) {
                if (knows[p]) {
                    queue.Enqueue(p);
                    visited.Add(p);
                }
            }

            while (queue.Count > 0) {
                int cur = queue.Dequeue();
                if (!adj.ContainsKey(cur)) continue;
                foreach (int nxt in adj[cur]) {
                    if (!visited.Contains(nxt)) {
                        visited.Add(nxt);
                        queue.Enqueue(nxt);
                    }
                }
            }

            foreach (int p in visited) knows[p] = true;
        }

        var result = new List<int>();
        for (int x = 0; x < n; x++)
            if (knows[x]) result.Add(x);

        return result;
    }
}
